src/hooks/useCreateMap.ts

```diff
@@ -0,0 +1,264 @@
+  const [user] = useAuthState(auth)
+  const cancelRef = useRef(false)
+  const uploadsRef = useRef(new Map())
+  const totalBytesRef = useRef(0)
+
+  const [totalFiles, setTotalFiles] = useState(0)
+  const [currentFile, setCurrentFile] = useState(0)
+  const [failedFiles, setFailedFiles] = useState<ImageFileType[]>([])
+  const [progress, setProgress] = useState(0)
+  const [loading, setLoading] = useState(false)
+  const [error, setError] = useState<Error | null>(null)
+
+  // useEffect(
+  //   // eslint-disable-next-line arrow-body-style
+  //   () => {
+  //     return () => {
+  //       // cancel uploads on unmount
+  //       cancelRef.current = true
+  //       for (const upload of uploadsRef.current.values()) {
+  //         if (typeof upload.cancel === 'function') {
+  //           // upload.cancel()
+  //         }
+  //       }
+  //     }
+  //   },
+  //   [],
+  // )
+
+  const createMapDoc = useCallback(
+    async (files: FileType[]) => {
+      if (!user) throw new Error('Not Authorized')
+
+      const metadata = getMetadata(files)
+      const mapsPath = `groups/${user.uid}/maps`
+
+      const mapDoc = await addDoc(collection(db, mapsPath), metadata)
+
+      return `${mapsPath}/${mapDoc.id}`
+    },
+    [user],
+  )
+
+  const uploadImage = useCallback(
+    (file: ImageFileType, current: number, total: number) => {
+      if (!user) throw new Error('Not Authorized')
+      // const emitter = new EventEmitter() as CancellableEventEmitterType
+      // emitter.cancel = () => {
+      //   emitter.removeAllListeners()
+      // }
+
+      const upload = { file, bytesTransferred: 0 }
+      uploadsRef.current.set(file.hashedName, upload)
+
+      console.log(`Uploading ${file.name}`)
+
+      const handleLastUpload = () => {
+        // If we on the last file, either on-success or on-error we want to unset loading state
+        if (current === total) {
+          console.log({ current, total })
+          setLoading(false)
+        }
+      }
+
+      const fileMeta = { contentType: 'image/jpeg' } // TODO: Support PNG
+      const storageRef = ref(storage, `images/${user.uid}/original/${file.hashedName || file.name}`)
+
+      // // <<! DEBUG CODE START !>>
+      // if (file.name.includes('3.png')) {
+      //   setFailedFiles((prevFailedFiles) => [...prevFailedFiles, file])
+      //   handleLastUpload()
+      //   return
+      // }
+      // // <<! END DEBUG CODE !>>
+
+      const uploadTask = uploadBytesResumable(storageRef, file.data, fileMeta)
+
+      uploadTask.on(
+        'state_changed',
+        (snapshot: UploadTaskSnapshot) => {
+          // Observe state change events such as progress, update the progress recorded.
+          upload.bytesTransferred = snapshot.bytesTransferred
+
+          updateProgress()
+        },
+        (uploadError: FirebaseError) => {
+          console.log({ error: uploadError })
+          setFailedFiles((prevFailedFiles) => [...prevFailedFiles, file])
```

Comment:
If a file fails, its being held in memory twice. Maybe it would be better to hold all the files in a ref, and just save the hashed names in `setFailedFiles` and then iterate through the `filesRef`, and only extract the failed files